deepquest/utils/callbacks.py

```diff
@@ -477,6 +479,17 @@ def evaluate(self, epoch, counter_name='epoch', logs={}):
             from keras_wrapper.cnn_model import saveModel
             saveModel(self.model_to_eval, epoch, store_iter=not self.eval_on_epochs)
 
+        # Update best epoch and score
+        if self.metric_check:
+            if metrics[self.metric_check] > self.best_score: #FIXME Pearson is maximised, but RMSE and MAE are minimised
+                self.best_score = metrics[self.metric_check]
+                self.best_epoch = epoch
+                src = self.save_path + '/' + counter_name + '_' + str(epoch) + '.h5'
+                tmp = self.save_path + '/' + 'tmp_best_' + counter_name + '.h5'
+                dst = self.save_path + '/' + 'best_' + counter_name + '.h5'
+                os.symlink(src, tmp)
+                os.rename(tmp, dst) # gets around overwriting/updating symlinks
+
```

Comment:
This is not necessary as the test whether the current epoch improves over the past one is done lines 1191-1224 in the `evaluate()` function of the `EarlyStopping` class:
```python
        # Check if the best score has been outperformed in the current epoch
        if current_score > self.best_score:
            self.best_epoch = epoch
            self.best_score = current_score
            self.threshold = threshold
            self.wait = 0
            if self.verbose > 0:
                if self.threshold is not None:
                    logging.info('---current best %s %s: %.3f threshold %.1f' % (self.check_split, self.metric_check,
                                                              current_score if not self.want_to_minimize
                                                              else -current_score, self.threshold))
                else:
                    logging.info('---current best %s %s: %.3f' % (self.check_split, self.metric_check,
                                                              current_score if not self.want_to_minimize
                                                              else -current_score))

        # Stop training if performance has not improved for self.patience epochs
        elif self.patience > 0:
            self.wait += 1
            logging.info('---bad counter: %d/%d' % (self.wait, self.patience))
            if self.wait >= self.patience:
                if self.verbose > 0:
                    if self.threshold is not None:
                        logging.info("---%s %d: early stopping. Best %s found at %s %d: %f threshold %.1f" % (
                             str(counter_name), epoch, self.metric_check, str(counter_name), self.best_epoch,
                             self.best_score if not self.want_to_minimize else -self.best_score, self.threshold))
                    else:
                        logging.info("---%s %d: early stopping. Best %s found at %s %d: %f" % (
                             str(counter_name), epoch, self.metric_check, str(counter_name), self.best_epoch,
                             self.best_score if not self.want_to_minimize else -self.best_score))

                self.model.stop_training = True
                exit(1)
```
The block above is what we see at the end of each training epoch.
My understanding is that we simply have to save the model here under `best_model.h5`, overwriting the previous one if it exists. Check the use of the `saveModel()` function for that, as in the `evaluate()` function of `EvalPerformance` (lines 475-478):
```python
        # Save the model
        if self.save_each_evaluation:
            from keras_wrapper.cnn_model import saveModel
            saveModel(self.model_to_eval, epoch, store_iter=not self.eval_on_epochs)
```
One issue I see with this function is that it doesn't take a "model name", therefore we need to find a workaround here. We could have our own local version of the function, but we depend a lot on `Keras_wrapper` therefore I would like to avoid copying a few functions here and there into deepQuest's codebase, such that if they change something one day (e.g. like having the architecture and the weights in a single file -- something they did when we were looking into that with @drj11 at the very begining) it shouldn't break and our goal is to keep the maintenance load as low as possible on that aspect.

Also, I would not use simlink as well, because if the parameter `SAVE_EACH_EVALUATION` is set to `False`, the file `epoch_X.h5` is not created. Same if the parameter `EPOCHS_FOR_SAVE` is bigger than 1.

Reply:
Perhaps I should have explained, sorry, the `EarlyStopping` callback used in training is specified in`keras_wrapper.cnn_model` (line 991, in `Model_Wrapper.__train`), imported on line 27 of the same file `from keras_wrapper.extra.callbacks import *` so we can't edit that callback. Hence why I put the code in with our `deepquest.utils.callbacks.evaluation()`
Correct me if I'm wrong or you have a solution.

To confuse matters, there's a version of EarlyStopping in `deepquest.utils.callbacks`, which is unused and I want to get rid of.

I will change the symlink to a model save.

Reply:
Have changed to using `saveModel` in 50cc2e6, certainly as it is currently written, we can hijack use the options of saveModel to give it a full path to save as follows:
```python
best_path = self.save_path + '/' + 'best_' + counter_name
from keras_wrapper.cnn_model import saveModel
saveModel(self.model_to_eval, epoch, store_iter=False, path=best_path, full_path=True)
```
which seems to do the trick.

Reply:
OK, that's good that we can use `path` to set the name. 
However, using `best_ + counter_name` will result in filling the disk space unnecessarily as it would simply duplicate each epoch until no improvement is observed. We should use `best_model.h5` as a "checkpoint" instead, overwriting the previous one if it exits.

Reply:
That is what happens here, the value of `counter_name` is (confusingly) the string `'epoch'` (not a digit), so the file written out goes to `trained_modes/model_name/best_epoch.h5` which naturally gets overwritten every time there's a better epoch.

Reply:
OK, my bad!

Reply:
No wukkas. I'll ask @drj11 if he thinks this approach is okay and we can merge?

Reply:
Can we make sure it works whatever the considered metrics or level of prediction, before we merge? I can see there is still a FIXME for instance.

Reply:
In the callback blocks that have been removed, there is a mechanism you may want to reuse:

```python
        if self.want_to_minimize:
            current_score = -current_score
```

Reply:
Yup, sorry got a bit over excited. Yes, this is probably the way I was going to do it. I see that there are a number of different metrics in `deepquest.utils.evaluation`:
- pearson, rmse and mae for sentences (& documents)
- Precision, Recall, f1_prod, and threshold for words
- Threshold, Precision, Recall, F-score, and F-score multi for phrases

`want_to_minimize` is slightly troubling in that it is only set when calling `EarlyStopping` during `Model_wrapper.__train()` and currently as:
```python
# Early stopper
if params.get('metric_check') is not None:
    callback_early_stop = EarlyStopping(self,
                                        patience=params['patience'],
                                        metric_check=params['metric_check'],
                                        want_to_minimize=True if 'TER' in params['metric_check'] else False,
                                        min_delta=params['min_delta'],
                                        eval_on_epochs=params['eval_on_epochs'],
                                        each_n_epochs=params['each_n_epochs'],
                                        start_eval_on_epoch=params['start_eval_on_epoch'])
    callbacks.append(callback_early_stop)
```
and so at present, is only set to True if the metric is `'TER'` (weirdly this is the same in the old codebase) so perhaps was only run with metrics to be maximised?

Reply:
RMSE and MAE are both error metrics hence to be minimised, although the metric of interest for training sent- and doc-level models is Pearson's r correlation (to be maximised).

Reply:
This is added by c2278fa